### PR TITLE
[6961] Add filter validation

### PR DESCRIPTION
**Pull request by stevehook**
### Context
We should be returning status 422 when a filter is invalid not just ignoring it.

Originally this was meant to be a spike but we've ended up with a 'complete' implementation.

### Changes proposed in this pull request
- Add a new `TraineeFilterParams` model that just does pre-validation of filter params before running the `TraineeFilter`.
- Example validation error response:
```
curl --request GET \
  --url localhost:5000/api/v0.1/trainees\?status=eating_breakfast\&academic_cycle=1066 \
  --header 'Authorization: Bearer xxxxxxxxxxxxxxxxxx'
{"message":"Validation failed: 2 errors prohibited this request being run","errors":{"status":["eating_breakfast is not a valid status"],"academic_cycle":["1066 is not a valid academic cycle year"]}}
```

### Guidance to review

- Is creating a separate 'validation model' sensible here?

### Important business

- [x] Does this PR introduce any PII fields that need to be overwritten or deleted in db/scripts/sanitise.sql?
- [x] Does this PR change the database schema? If so, have you updated the config/analytics.yml file and considered whether you need to send 'import_entity' events?
- [x] Do we need to send any updates to DQT as part of the work in this PR?
- [x] Does this PR need an ADR?

NB: Please notify the #twd_data_insights team and ask for a review if new fields are being added to analytics.yml

### CODE REVIEW

File: spec/models/api/trainee_filter_params_spec.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+describe Api::TraineeFilterParams do
+  describe "since validation" do
+    it "validates since date" do
```
Comment (Reyko) [inline spec/models/api/trainee_filter_params_spec.rb:7]: I would suggest to add an expectation when the date throws a `Date::Error` according to the code

File: spec/requests/api/v0.1/get_trainees_spec.rb [hunk 2]
```diff
@@ -36,6 +36,16 @@
...
     end
+
+    it "returns 422 if academic cycle filter is invalid" do
+      get(
+        "/api/v0.1/trainees",
+        headers: { Authorization: "Bearer #{token}" },
+        params: { academic_cycle: "not-a-number" },
+      )
+
+      expect(response).to have_http_status(:unprocessable_entity)
```
Comment (Reyko) [inline spec/requests/api/v0.1/get_trainees_spec.rb:47]: Maybe, add an expectation for the response body with the error message

**Review by Reyko — COMMENTED** [review]
Nice one! Left some suggestions for the tests 👍

File: spec/models/api/trainee_filter_params_spec.rb (hunk 1 above)
Comment (Reyko) [reply spec/models/api/trainee_filter_params_spec.rb]: Yes my bad! Nice

File: app/models/api/trainee_filter_params.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    validate :check_academic_cycle
+    validates :sort_order, inclusion: %w[desc asc], allow_blank: true
+
+  private
+
+    def check_statuses
+      return if status.blank?
+
+      (status.is_a?(Array) ? status : [status]).each do |status_value|
+        errors.add(:status, "#{status_value} is not a valid status") unless Trainee.states.include?(status_value)
```
Comment (tomtrentham) [inline app/models/api/trainee_filter_params.rb:22]: Do we need to be checking against `TraineeFilter::STATUSES` here rather than `Trainee.states`?

**Review by tomtrentham — APPROVED** [review]
Nice one. `status: :approved`

**Pull request merged**